src/intrusive_pointer.rs

```diff
@@ -99,6 +99,38 @@ unsafe impl<T: ?Sized> IntrusivePointer<T> for Arc<T> {
     }
 }
 
+/// Creates an `IntrusivePointer` from a raw pointer
+///
+/// This method is only safe to call if the raw pointer is known to be
+/// managed by the provided `IntrusivePointer` type.
+pub(crate) unsafe fn clone_pointer_from_raw<P: IntrusivePointer<T> + Clone, T: ?Sized>(
+    pointer: *const T,
+) -> P {
+    /// Guard which converts an `IntrusivePointer` back into its raw version
+    /// when it gets dropped. This makes sure we also perform a full
+    /// `from_raw` and `into_raw` round trip - even in the case of panics.
+    struct PointerGuard<P: IntrusivePointer<T>, T: ?Sized> {
+        pointer: Option<P>,
+        _phantom: core::marker::PhantomData<T>,
+    }
+
+    impl<P: IntrusivePointer<T>, T: ?Sized> Drop for PointerGuard<P, T> {
+        fn drop(&mut self) {
+            if let Some(pointer) = self.pointer.take() {
+                // Prevent shared pointers from being released by converting them
+                // back into the raw pointers
+                let _ = pointer.into_raw();
+            }
+        }
+    }
+
+    let holder = PointerGuard {
+        pointer: Some(P::from_raw(pointer)),
+        _phantom: core::marker::PhantomData,
+    };
+    holder.pointer.as_ref().map(|p| p.clone()).unwrap()
```

Comment:
Rather than using `as_ref`, just dereference the pointer directly and specify that it can't be called with a null pointer (it's an unsafe function anyways). That way you don't need to deal with the `Option` mess.

Reply:
The `Option` is not there to allow to pass non-null pointers. It is there because we can't just call `pointer.into_raw()` in the destructor. That would require moving the field out of the struct, which is not allowed - aparently even not in the destructor.

We could alternatively store the pointer in a `ManualDrop`, and pass a `.clone()` of it to `into_raw()`. But then we are basically just back to the original implementation.

Which means I do not really know how to do this more elegant.

Reply:
Ah I see, I misunderstood the code (I thought you were calling `as_ref` on the pointer).

However could you please just use unwrap directly instead of `.map` to make things clearer:
```suggestion
    holder.pointer.as_ref().unwrap().clone()
```

Same thing in the `Drop`, just use `self.pointer.take().unwrap()`.

Reply:
Done